src/Compilers/Core/AnalyzerDriver/DiagnosticAnalysisContextHelpers.cs

```diff
@@ -25,12 +25,17 @@ internal static void VerifyArguments<TContext>(Action<TContext> action, Immutabl
             VerifySyntaxKinds(syntaxKinds);
         }
 
-        internal static void VerifyArguments(Diagnostic diagnostic)
+        internal static void VerifyArguments(Diagnostic diagnostic, Func<Diagnostic, bool> isSupportedDiagnostic)
         {
             if (diagnostic == null)
             {
                 throw new ArgumentNullException(nameof(diagnostic));
             }
+
+            if (!isSupportedDiagnostic(diagnostic))
+            {
+                throw new ArgumentException(string.Format(AnalyzerDriverResources.UnsupportedDiagnosticReported, diagnostic.Id));
```

Comment:
Consider passing `nameof(diagnostic)` as the second argument to the constructor.